File: frontend/src/api/campaign.ts
import { api } from './client'

export interface CampaignTarget {
  id: number
  apk_filename: string
  analysis_id: number | null
  status: string
  error: string | null
}

export interface CampaignOut {
  id: number
  created_at: string
  name: string
  description: string | null
  platform: string
  status: string
  targets: CampaignTarget[]
}

export interface CampaignSummary {
  id: number
  created_at: string
  name: string
  platform: string
  status: string
  total: number
  complete: number
  failed: number
}

export const campaignApi = {
  create: (name: string, description?: string, platform = 'android') =>
    api.post<CampaignOut>('/campaigns', { name, description, platform }).then((r) => r.data),

  list: () => api.get<CampaignSummary[]>('/campaigns').then((r) => r.data),

  get: (id: number) => api.get<CampaignOut>(`/campaigns/${id}`).then((r) => r.data),

  addTarget: (campaignId: number, file: File) => {
    const form = new FormData()
    form.append('file', file)
    return api.post<CampaignOut>(`/campaigns/${campaignId}/targets`, form, {
      headers: { 'Content-Type': 'multipart/form-data' },
    }).then((r) => r.data)
  },

  run: (id: number) => api.post<CampaignOut>(`/campaigns/${id}/run`).then((r) => r.data),

  delete: (id: number) => api.delete(`/campaigns/${id}`),
}

File: backend/schemas/campaign.py
from datetime import datetime

from pydantic import BaseModel


class CampaignCreate(BaseModel):
    name: str
    description: str | None = None
    platform: str = "android"   # android | ios | mixed


class CampaignTargetOut(BaseModel):
    id: int
    apk_filename: str
    analysis_id: int | None
    status: str
    error: str | None

    model_config = {"from_attributes": True}


class CampaignOut(BaseModel):
    id: int
    created_at: datetime
    name: str
    description: str | None
    platform: str
    status: str
    targets: list[CampaignTargetOut] = []

    model_config = {"from_attributes": True}


class CampaignSummary(BaseModel):
    id: int
    created_at: datetime
    name: str
    platform: str
    status: str
    total: int = 0
    complete: int = 0
    failed: int = 0

    model_config = {"from_attributes": True}
